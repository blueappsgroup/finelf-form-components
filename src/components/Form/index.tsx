import React, { FC, ReactElement, useMemo, useState } from 'react'
import { Formik, Form } from 'formik'
import styled from 'styled-components'

import { ThemeProvider } from '../../consts/theme'
import { device } from '../../consts/sizes'
import { FormProps, FormValuesType } from '../../types'
import RedirectPage from '../RedirectPage'
import {
  FormContext,
  getFormValuesFromCache,
  resetFormValueCache,
  handleSendDataToApi,
  getFieldsValuesFromUrl,
} from '../../utils'
import { formStatuses } from '../../consts/form'

const StyledForm = styled(Form)`
  display: flex;
  flex-direction: column;
  max-width: ${(props): string => props.theme.formMaxWidth};
  justify-self: center;
  background: ${(props): string => props.theme.formBgColor};
  font-family: ${(props): string => props.theme.fontFamily};
  border-radius: 6px;
  box-shadow: ${(props): string => props.theme.formBoxShadow};
`

const FormWrapper: FC<FormProps> = ({
  children,
  onSubmit,
  customTheme,
  id,
  stepsLength,
  stepsTitles,
  hasRedirect,
  redirectUrl,
  timeToRedirect,
  redirectHeaderText,
  logoImg,
  redirectMainImg,
  redirectBgImg,
  sendDataToApi,
  apiUrl,
  transactionName,
  propertyNamesFromUrl,
  dataWithUserAgent,
}) => {
  const trasationIdValue =
    transactionName &&
    new URLSearchParams(window.location.search).get(transactionName)
  const intialValuesFromUrl = useMemo(
    () =>
      (propertyNamesFromUrl &&
        propertyNamesFromUrl.length > 0 &&
        getFieldsValuesFromUrl(propertyNamesFromUrl)) ||
      {},
    [propertyNamesFromUrl]
  )
  const [redirectUrlPath, setRedirectUrlPath] = useState(redirectUrl)
  const [initialValues, setInitialValues] = useState({
    ...intialValuesFromUrl,
    ...getFormValuesFromCache(id),
    // eslint-disable-next-line @typescript-eslint/camelcase
    trasaction_id: trasationIdValue,
  })
  const [currentStep, setCurrentStep] = useState(0)
  const [fieldsForSkip, setFieldsForSkip] = useState<string[]>([])
  const [errorFromApi, setErrorFromApi] = useState<boolean>(false)
  const shouldRedirect = !errorFromApi && hasRedirect
  const addFieldForSkip = (key: string): void =>
    setFieldsForSkip([...fieldsForSkip, key])

  const handleSubmit = async (
    values: FormValuesType,
    props: {
      resetForm: () => void
      setStatus: (status: string) => void
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): Promise<any> => {
    if (onSubmit) {
      onSubmit(values, props)
    }

    try {
      if (sendDataToApi && apiUrl) {
        const response = await handleSendDataToApi(
          values,
          apiUrl,
          id,
          fieldsForSkip,
          dataWithUserAgent
        )
        const {
          redirectUrl: urlFromApi,
          status: statusFromApi,
        } = await response.json()

        if (statusFromApi === false) {
          setErrorFromApi(true)
          return
        }

        urlFromApi && setRedirectUrlPath(urlFromApi)
      }

      props.resetForm()
      props.setStatus(formStatuses.submited)
    } catch (e) {
      props.setStatus(formStatuses.error)
    }
  }
  const handleReset = (): void => {
    resetFormValueCache(id)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setInitialValues({} as any)
  }

  const prevStep: Function = () => setCurrentStep(currentStep - 1)

  const nextStep: Function = () => setCurrentStep(currentStep + 1)

  return (
    <FormContext.Provider
      value={{
        id,
        apiUrl,
        stepsLength,
        currentStep,
        stepsTitleList: stepsTitles,
        nextStep,
        prevStep,
        fieldsForSkip,
        addFieldForSkip,
        errorFromApi,
      }}
    >
      <ThemeProvider customTheme={{ ...customTheme }}>
        <Formik
          enableReinitialize
          initialValues={initialValues}
          onSubmit={handleSubmit}
          onReset={handleReset}
        >
          {(props): ReactElement =>
            (shouldRedirect && props.status === formStatuses.submited && (
              <RedirectPage
                redirectUrl={redirectUrlPath}
                backgroundImage={redirectBgImg}
                logoImg={logoImg}
                headerText={redirectHeaderText}
                timeToRedirect={timeToRedirect}
                mainImg={redirectMainImg}
              />
            )) || (
              <StyledForm id={id}>
                {children}
              </StyledForm>
            )
          }
        </Formik>
      </ThemeProvider>
    </FormContext.Provider>
  )
}

export default FormWrapper
