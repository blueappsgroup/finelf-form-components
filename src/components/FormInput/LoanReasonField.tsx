/* eslint-disable @typescript-eslint/camelcase */
import { Field } from 'formik'
import React, { ReactElement } from 'react'
import { SelectFieldProps } from '../../types'

import { BaseSelectField } from './base'
import { validateSelect } from './validate'

const options = {
  bills: 'rachunki',
  other_loans: 'inne pożyczki',
  vacation: 'wakacje',
  education: 'edukacja',
  car: 'samochód',
  medical_expenses: 'wydatki zdrowotne',
  renovation: 'renowacja',
  electronics: 'elektronika',
  furniture: 'meble',
  other: 'inne',
  investment: 'inwestycja',
  tax_payment: 'spłata podatków',
}

const LoanReasonField: (props: SelectFieldProps) => ReactElement = ({
  id,
  label,
  name = 'loan_reason',
  required = false,
  validate,
  showError,
}) => (
  <Field
    required={required}
    showError={showError}
    id={id}
    name={name}
    component={BaseSelectField}
    label={label}
    type="select"
    options={options}
    validate={
      (validate && validate(required)) || validateSelect(required, options)
    }
  />
)

export default LoanReasonField
