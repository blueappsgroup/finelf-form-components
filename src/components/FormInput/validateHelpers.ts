export const isEmpty = (value: string): boolean => !value

export const isNotValidLength = (value: string, from = 2, to = 100): boolean =>
  value.length < from || value.length > to

export const isNotValidPhoneNumber = (value: string): boolean =>
  !/^(45|50|51|53|57|60|66|69|72|73|78|79|88)[0-9]{7}$/.test(value)

export const isNotValidEmail = (value: string): boolean =>
  !/[a-zA-Z0-9_.+-]+@[a-zAZ0-9-]+\.[a-zA-Z0-9-.]+$/.test(value)

export const isNotValidOtherLoanAmount = (value: string): boolean =>
  !/^[0-9]{1,10}$/.test(value)
