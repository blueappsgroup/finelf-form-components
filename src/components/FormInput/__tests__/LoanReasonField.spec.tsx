import * as React from 'react'
import { render, fireEvent, act } from '@testing-library/react'

import Form from '../../Form'
import LoanReasonField from '../LoanReasonField'
import { isNotValidOption } from '../validateHelpers'

describe('<LoanReasonField />', () => {
  const onSubmit = jest.fn()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const setupWrapper: (config: any) => any = ({
    formId = 'testForm',
    selectId = 'loanReasonField',
    selectName = 'name',
    required = false,
    ...rest
  }) => {
    const wrapper = render(
      <Form id={formId} onSubmit={onSubmit}>
        <LoanReasonField
          id={selectId}
          name={selectName}
          required={required}
          {...rest}
        />
      </Form>
    )
    const { container } = wrapper

    const select = container.querySelector(`[name="${selectName}"]`)

    return {
      select,
      ...wrapper,
    }
  }

  it('matches snapshot', () => {
    const { select } = setupWrapper({})

    expect(select).toBeTruthy()
    expect(select).toMatchSnapshot()
  })

  it('matches snapshot with label', () => {
    const wrapper = setupWrapper({ label: 'Test' })
    const { container } = wrapper

    expect(wrapper.getByText('Test')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('matches snapshot with error', async () => {
    const wrapper = setupWrapper({
      label: 'Test',
      showError: true,
      required: true,
      error: 'true',
    })
    const { container, select } = wrapper

    await act(async () => {
      fireEvent.focus(select, { target: { value: 'bills' } })
    })

    await act(async () => {
      fireEvent.blur(select, { target: { value: 'bills' } })
    })

    expect(wrapper.getByText('To pole jest wymagane')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it('changes select field', async () => {
    const { select } = setupWrapper({})

    await act(async () => {
      fireEvent.change(select, { target: { value: 'investment' } })
    })

    expect(select.value).toBe('investment')
  })

  it('select field onBlur save value to sessionStorage', async () => {
    const { select } = setupWrapper({})

    await act(async () => {
      fireEvent.change(select, { target: { value: 'investment' } })
    })

    await act(async () => {
      fireEvent.blur(select, { target: { value: 'investment' } })
    })

    expect(
      JSON.parse(global.window.sessionStorage.getItem('form-testForm')).name
    ).toBe('investment')
    expect(select.value).toBe('investment')
  })

  it('select field with custom validation', async () => {
    const customValidate = () => (value: string): string => {
      /* eslint-disable @typescript-eslint/camelcase */
      if (
        isNotValidOption(value, {
          bills: 'rachunki',
          other_loans: 'inne pożyczki',
          vacation: 'wakacje',
          education: 'edukacja',
          car: 'samochód',
          medical_expenses: 'wydatki zdrowotne',
          renovation: 'renowacja',
          electronics: 'elektronika',
          furniture: 'meble',
          other: 'inne',
          investment: 'inwestycja',
          tax_payment: 'spłata podatków',
        })
      ) {
        return 'Podane dane są nieprawidłowe'
      }
    }
    const wrapper = setupWrapper({
      validate: customValidate,
      showError: true,
    })
    const { select, container } = wrapper

    await act(async () => {
      fireEvent.change(select, { target: { value: 'test' } })
    })

    await act(async () => {
      fireEvent.blur(select, { target: { value: 'test' } })
    })

    expect(container).toMatchSnapshot()
    expect(wrapper.getByText('Podane dane są nieprawidłowe')).toBeTruthy()
  })
})
