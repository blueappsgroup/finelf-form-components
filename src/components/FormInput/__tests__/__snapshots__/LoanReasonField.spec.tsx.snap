// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoanReasonField /> matches snapshot 1`] = `
<select
  class="sc-iemXMA ezYUqe"
  id="loanReasonField"
  name="name"
  type="text"
>
  <option
    class="sc-hHEjAm hmYdyr"
    disabled=""
    value="select"
  >
    wybierz 
  </option>
  <option
    value="bills"
  >
    rachunki
  </option>
  <option
    value="other_loans"
  >
    inne pożyczki
  </option>
  <option
    value="vacation"
  >
    wakacje
  </option>
  <option
    value="education"
  >
    edukacja
  </option>
  <option
    value="car"
  >
    samochód
  </option>
  <option
    value="medical_expenses"
  >
    wydatki zdrowotne
  </option>
  <option
    value="renovation"
  >
    renowacja
  </option>
  <option
    value="electronics"
  >
    elektronika
  </option>
  <option
    value="furniture"
  >
    meble
  </option>
  <option
    value="other"
  >
    inne
  </option>
  <option
    value="investment"
  >
    inwestycja
  </option>
  <option
    value="tax_payment"
  >
    spłata podatków
  </option>
</select>
`;

exports[`<LoanReasonField /> matches snapshot with error 1`] = `
<div>
  <form
    action="#"
    class="sc-iCoHVE JvWob"
    id="testForm"
  >
    <div
      class="sc-kEqYlL dXeAFH"
    >
      <label
        for="name"
      >
        Test*
      </label>
      <div
        class="sc-bkbjAj gnGQfC"
      >
        <select
          class="sc-iemXMA varr"
          id="loanReasonField"
          label="Test"
          name="name"
          required=""
          type="text"
        >
          <option
            class="sc-hHEjAm hmYdyr"
            disabled=""
            value="select"
          >
            wybierz *
          </option>
          <option
            value="bills"
          >
            rachunki
          </option>
          <option
            value="other_loans"
          >
            inne pożyczki
          </option>
          <option
            value="vacation"
          >
            wakacje
          </option>
          <option
            value="education"
          >
            edukacja
          </option>
          <option
            value="car"
          >
            samochód
          </option>
          <option
            value="medical_expenses"
          >
            wydatki zdrowotne
          </option>
          <option
            value="renovation"
          >
            renowacja
          </option>
          <option
            value="electronics"
          >
            elektronika
          </option>
          <option
            value="furniture"
          >
            meble
          </option>
          <option
            value="other"
          >
            inne
          </option>
          <option
            value="investment"
          >
            inwestycja
          </option>
          <option
            value="tax_payment"
          >
            spłata podatków
          </option>
        </select>
        <span
          class="sc-dIvqjp gGSLdt"
        />
      </div>
      <span
        class="sc-ksluoS cNvzFC"
      >
        To pole jest wymagane
      </span>
    </div>
  </form>
</div>
`;

exports[`<LoanReasonField /> matches snapshot with label 1`] = `
<div>
  <form
    action="#"
    class="sc-iCoHVE JvWob"
    id="testForm"
  >
    <div
      class="sc-kEqYlL dXeAFH"
    >
      <label
        for="name"
      >
        Test
      </label>
      <div
        class="sc-bkbjAj gnGQfC"
      >
        <select
          class="sc-iemXMA ezYUqe"
          id="loanReasonField"
          label="Test"
          name="name"
          type="text"
        >
          <option
            class="sc-hHEjAm hmYdyr"
            disabled=""
            value="select"
          >
            wybierz 
          </option>
          <option
            value="bills"
          >
            rachunki
          </option>
          <option
            value="other_loans"
          >
            inne pożyczki
          </option>
          <option
            value="vacation"
          >
            wakacje
          </option>
          <option
            value="education"
          >
            edukacja
          </option>
          <option
            value="car"
          >
            samochód
          </option>
          <option
            value="medical_expenses"
          >
            wydatki zdrowotne
          </option>
          <option
            value="renovation"
          >
            renowacja
          </option>
          <option
            value="electronics"
          >
            elektronika
          </option>
          <option
            value="furniture"
          >
            meble
          </option>
          <option
            value="other"
          >
            inne
          </option>
          <option
            value="investment"
          >
            inwestycja
          </option>
          <option
            value="tax_payment"
          >
            spłata podatków
          </option>
        </select>
        <span
          class="sc-dIvqjp gGSLdt"
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`<LoanReasonField /> select field with custom validation 1`] = `
<div>
  <form
    action="#"
    class="sc-iCoHVE JvWob"
    id="testForm"
  >
    <div
      class="sc-kEqYlL dXeAFH"
    >
      <div
        class="sc-bkbjAj gnGQfC"
      >
        <select
          class="sc-iemXMA varr"
          id="loanReasonField"
          name="name"
          type="text"
        >
          <option
            class="sc-hHEjAm hmYdyr"
            disabled=""
            value="select"
          >
            wybierz 
          </option>
          <option
            value="bills"
          >
            rachunki
          </option>
          <option
            value="other_loans"
          >
            inne pożyczki
          </option>
          <option
            value="vacation"
          >
            wakacje
          </option>
          <option
            value="education"
          >
            edukacja
          </option>
          <option
            value="car"
          >
            samochód
          </option>
          <option
            value="medical_expenses"
          >
            wydatki zdrowotne
          </option>
          <option
            value="renovation"
          >
            renowacja
          </option>
          <option
            value="electronics"
          >
            elektronika
          </option>
          <option
            value="furniture"
          >
            meble
          </option>
          <option
            value="other"
          >
            inne
          </option>
          <option
            value="investment"
          >
            inwestycja
          </option>
          <option
            value="tax_payment"
          >
            spłata podatków
          </option>
        </select>
        <span
          class="sc-dIvqjp gGSLdt"
        />
      </div>
      <span
        class="sc-ksluoS cNvzFC"
      >
        Podane dane są nieprawidłowe
      </span>
    </div>
  </form>
</div>
`;
